velox/functions/sparksql/specialforms/GetStructField.cpp

```diff
@@ -0,0 +1,109 @@
+/*
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "velox/functions/sparksql/specialforms/GetStructField.h"
+#include "velox/expression/ConstantExpr.h"
+
+namespace facebook::velox::functions::sparksql {
+namespace {
+
+// Returns the value of nested field in the input ``args``.
+// The input ``args`` must be of row type and nested complex type is allowed.
+// The field position is specified by ``ordinal``.
+// If ``ordinal`` is negative or greater than the children size of input,
+// exception is thrown.
+class GetStructFieldFunction : public exec::VectorFunction {
+ public:
+  GetStructFieldFunction(int32_t ordinal) : ordinal_(ordinal) {}
+
+  void apply(
+      const SelectivityVector& rows,
+      std::vector<VectorPtr>& args,
+      const TypePtr& resultType,
+      exec::EvalCtx& context,
+      VectorPtr& result) const override {
+    exec::LocalDecodedVector decoded(context, *args[0], rows);
+    auto rowData = decoded->base()->as<RowVector>();
+    VELOX_USER_CHECK_GE(
+        ordinal_, 0, "Invalid ordinal. Should be greater than 0.");
+    VELOX_USER_CHECK_LT(
+        ordinal_,
+        rowData->childrenSize(),
+        "Invalid ordinal. Should be small than the children size of input row vector {}.",
```

Comment:
nit: perhaps also include the ordinal value in the error message.

Reply:
We don't need to include the ordinal and rowData->childrenSize(), macro VELOX_USER_CHECK_LT will print these information.

Reply:
The details appears not included in the error message, could you confirm?

https://github.com/facebookincubator/velox/pull/12166/files#diff-2b2cd7697bed01bbac99cd4c2138c5817fbad445a8358416ccd06b18dfd2975eR115

Reply:
yes, I'v checked that the full error message is `'(4 vs. 3) Invalid ordinal. Should be smaller than the children size of input row vector.`

`VELOX_ASSERT_THROW` here only check if the full error message `contains` specific error message.